src/Core/Checkout/Payment/DataAbstractionLayer/PaymentMethodIndexer.php

```diff
@@ -67,7 +69,11 @@ public function handle(EntityIndexingMessage $message): void
             return;
         }
 
-        $this->distinguishableNameGenerator->generateDistinguishablePaymentNames($message->getContext());
+        // Create a new context with disabled-indexing state, because DAL is used inside to upsert payment methods.
+        $newContext = Context::createFrom($message->getContext());
+        // Apparently the new context loses the states of the old one during creation. So the old states get added back.
+        $newContext->addState(EntityIndexerRegistry::DISABLE_INDEXING, ...$message->getContext()->getStates());
```

Comment:
If you remove the `EntityIndexerRegistry::DISABLE_INDEXING` here or just use the context from the message directly in line 76, you'll see the test failing.